src/java/org/apache/cassandra/io/tries/Walker.java

```diff
@@ -0,0 +1,381 @@
+     * On return the walker is positioned at the longest prefix that matches the input (with or without payload), and
+     * max(lesserBranch) is the immediate lesser neighbour.
+     *
+     * @return the first unmatched byte of the key, may be {@link ByteSource#END_OF_STREAM}
+     */
+    public int followWithLesser(ByteComparable key)
+    {
+        lesserBranch = -1;
+
+        ByteSource stream = key.asComparableBytes(BYTE_COMPARABLE_VERSION);
+        go(root);
+        while (true)
+        {
+            int b = stream.next();
+            int searchIndex = search(b);
+
+            lesserBranch = lesserTransition(searchIndex, lesserBranch);
+
+            if (searchIndex < 0)
+                return b;
+
+            go(transition(searchIndex));
+        }
+    }
+
+
+    /**
+     * Takes a prefix of the given key. The prefix is in the sense of a separator key match, i.e. it is only
+     * understood as valid if there are no greater entries in the trie (e.g. data at 'a' is ignored if 'ab' or 'abba'
+     * is in the trie when looking for 'abc' or 'ac', but accepted when looking for 'aa').
+     * In order to not have to go back to data that may have exited cache, payloads are extracted when the node is
+     * visited (instead of saving the node's position), which requires an extractor to be passed as parameter.
+     */
+    @SuppressWarnings("unchecked")
+    public <RESULT> RESULT prefix(ByteComparable key, Extractor<RESULT, VALUE> extractor)
+    {
+        RESULT payload = null;
+
+        ByteSource stream = key.asComparableBytes(BYTE_COMPARABLE_VERSION);
+        go(root);
+        while (true)
+        {
+            int b = stream.next();
+            int childIndex = search(b);
+
+            if (childIndex > 0)
+                payload = null;
+            else
+            {
+                int payloadBits = payloadFlags();
+                if (payloadBits > 0)
+                    payload = extractor.extract((VALUE) this, payloadPosition(), payloadBits);
+                if (childIndex < 0)
+                    return payload;
+            }
+
+            go(transition(childIndex));
+        }
+    }
+
+    /**
+     * Follows the trie for a given key, taking a prefix (in the sense above) and searching for neighboring values.
+     * On return min(greaterBranch) and max(lesserBranch) are the immediate non-prefix neighbours for the sought value.
+     * <p>
+     * Note: in a separator trie the closest smaller neighbour can be another prefix of the given key. This method
+     * does not take that into account. E.g. if trie contains "abba", "as" and "ask", looking for "asking" will find
+     * "ask" as the match, but max(lesserBranch) will point to "abba" instead of the correct "as". This problem can
+     * only occur if there is a valid prefix match.
+     */
+    @SuppressWarnings("unchecked")
+    public <RESULT> RESULT prefixAndNeighbours(ByteComparable key, Extractor<RESULT, VALUE> extractor)
```

Comment:
I guess we have `hasChildren()`, `prefixAndNeighbours()`, `followWithGreater()`, and `followWithLesser()` that are unused for SAI, but probably used later for the trie primary index. At this point I'm just drowning with the size of the patch, so any opportunity to prune would be nice. WDYT?